camel-core/src/main/java/org/apache/camel/ComponentVerifier.java

```diff
@@ -21,118 +21,321 @@
+         * recommended to stick to the predefined standard codes
+         *
+         * @return the general error code.
          */
-        String getCode();
+        Code getCode();
 
         /**
+         * A human readable description of the error in plain english
+         *
          * @return the error description (if available)
          */
         String getDescription();
 
         /**
-         * @return the parameters in error
+         * A set of input parameter names which fails the verification. These are keys to the parameter provided
+         * to {@link #verify(Scope, Map)}.
+         *
+         * @return the parameter names which are malformed and caused the failure of the validation
          */
-        Set<String> getParameters();
+        Set<String> getParameterKeys();
 
         /**
-         * @return a number of key/value pair with additional information related to the validation.
+         * Details about the failed verification. The keys can be either predefined values
+         * ({@link ExceptionAttribute}, {@link HttpAttribute}, {@link GroupAttribute}) or it can be free-form
+         * custom keys specific to a component. The standard attributes are defined as enums in all uppercase (with
+         * underscore as separator), custom attributes are supposed to be in all lower case (also with underscores
+         * as separators)
+         *
+         * @return a number of key/value pair with additional information related to the verification.
          */
-        Map<String, Object> getAttributes();
-    }
+        Map<Attribute, Object> getDetails();
 
-    /**
-     * Represent a validation Result.
-     */
-    interface Result extends Serializable {
-        enum Status {
-            OK,
-            ERROR,
-            UNSUPPORTED
+        /**
+         * Get a single detail for a given attribute
+         *
+         * @param attribute the attribute to lookup
+         * @return the detail value or null if no such attribute exists
+         */
+        default Object getDetail(Attribute attribute) {
+            Map<Attribute, Object> details = getDetails();
+            if (details != null) {
+                for (Map.Entry<Attribute, Object> entry : details.entrySet()) {
+                    if (ObjectHelper.equal(entry.getKey(), attribute)) {
+                        return entry.getValue();
+                    }
+                }
+            }
+            return null;
         }
 
         /**
-         * @return the scope against which the parameters have been validated.
+         * Get a single detail for a given attribute
+         *
+         * @param attribute the attribute to lookup
+         * @return the detail value or null if no such attribute exists
          */
-        Scope getScope();
+        default Object getDetail(String attribute) {
+            return getDetail(new ResultErrorBuilder.ErrorAttribute(attribute));
```

Comment:
I wonder whether we should not move `asAttribute()` and `asCode()` as `asAttribute()` and `asCode()` directly to the VerifierError as static methods. The reason is, that this method here introduces a circular dependencies to ResultErrorBuilder which could be avoided that way.

wdyt, @lburgazzoli  ?